File: app/src/main/java/com/islam/basepropject/project_base/base/adapters/BaseViewHolder.java
package com.islam.basepropject.project_base.base.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import com.islam.basepropject.R;

public abstract class BaseViewHolder<T> extends RecyclerView.ViewHolder {

    public BaseViewHolder(ViewGroup viewGroup,int layoutId) {
        super(LayoutInflater.from(viewGroup.getContext()).inflate(layoutId,viewGroup,false));
    }

    public abstract void onBind(T item);
}

File: app/src/main/java/com/islam/basepropject/project_base/base/adapters/BaseViewHolder.kt
package com.islam.basepropject.project_base.base.adapters

import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView


abstract class BaseViewHolder<T>(viewGroup: ViewGroup, layoutId: Int) : RecyclerView.ViewHolder((LayoutInflater.from(viewGroup.context).inflate(layoutId, viewGroup, false))) {

    init {
        super.itemView.isClickable =true
        super.itemView.addRippleEffect()
    }

    abstract fun onBind(item: T)


    private fun View.addRippleEffect() = with(TypedValue()) {
        context.theme.resolveAttribute(android.R.attr.selectableItemBackground, this, true)
        setBackgroundResource(resourceId)
    }
}